### ISSUE TITLE
Foreign keys can be declared with CASCADE

### ISSUE THREAD

**Issue by pgulutzan** [tarantool/tarantool#3475]
ON UPDATE|DELETE CASCADE etc. are accepted, but they do not work.
They should cause syntax errors until support exists.
box.sql.execute("CREATE TABLE te16 (s1 INT PRIMARY KEY, s2 INT, FOREIGN KEY (s2) REFERENCES te16 (s1) ON UPDATE CASCADE);")
Result = no error

**Comment by Korablev77** [comment]
@pgulutzan 
Could you please provide example of non-working CASCADE clause?
I have checked out from fresh 2.0:
```
CREATE TABLE t1(id PRIMARY KEY, a UNIQUE, b);
CREATE TABLE t2(id PRIMARY KEY, a REFERENCES t1(a) ON DELETE CASCADE);
INSERT INTO t1 VALUES (1, 2, 3);
INSERT INTO t2 VALUES (1, 2), (2, 2);
DELETE FROM t1;
SELECT * FROM t2;
---
- []
...

```
Thus, at least in simple case - they seem to work.

**Comment by pgulutzan** [comment]
You are right. CASCADE works now with simple cases, and
even with some complicated cases (I tried to find major
problems and I failed). However, I see four minor problems.
If you decide they are not bugs, then please close this report.

#1: If I have ON UPDATE CASCADE, and I set referenced-key value to NULL,
there's no cascade, there's an error. Actually making a reference to
a nullable column is bad anyway, so you can say "this is deliberate".
DROP TABLE t2;
DROP TABLE t1;
CREATE TABLE t1(a INT PRIMARY KEY, b INT UNIQUE);
CREATE TABLE t2(a INT PRIMARY KEY, b INT REFERENCES t1(b) ON UPDATE CASCADE);
INSERT INTO t1 VALUES (1, 2);
INSERT INTO t2 VALUES (1, 2);
UPDATE t1 SET b = NULL;

#2: REPLACE causes "ON DELETE CASCADE" action. Actually that is the
correct thing to do according to our definition (that REPLACE can
be equivalent to a DELETE and an INSERT). And that is consistent
with the way that REPLACE affects triggers (both the DELETE trigger
and the INSERT trigger are activated). My only worry is: I forget,
was there a conscious decision that SQL REPLACE will do this?
If there was, you can say "this is deliberate".
DROP TABLE t3;
DROP TABLE t2;
DROP TABLE t1;
CREATE TABLE t1(a INT PRIMARY KEY, b INT UNIQUE);
CREATE TABLE t2(a INT PRIMARY KEY, b INT REFERENCES t1(b) ON UPDATE CASCADE ON DELETE CASCADE);
INSERT INTO t1 VALUES (1, 1);
INSERT INTO t2 VALUES (1, 1);
INSERT INTO t2 VALUES (2, 1);
CREATE TABLE t3 (a INT PRIMARY KEY);
INSERT INTO t3 VALUES (0);
CREATE TRIGGER t1a BEFORE INSERT ON t1 FOR EACH ROW BEGIN UPDATE t3 SET a = a + 1; END;
CREATE TRIGGER t1b BEFORE DELETE ON t1 FOR EACH ROW BEGIN UPDATE t3 SET a = a + 1; END;
REPLACE INTO t1 VALUES (1, null);
-- The following select returns 0 rows
SELECT * FROM t2;
-- The following select shows that t3 was updated twice
SELECT * FROM t3;

#3: I can repeat ON UPDATE CASCADE without seeing a syntax error.
I admit, though, that this sort of "error" is so common that it
can be considered normal behaviour. So you can say "this is deliberate".
DROP TABLE t2;
DROP TABLE t1;
CREATE TABLE t1(a INT PRIMARY KEY, b INT UNIQUE);
CREATE TABLE t2(a INT PRIMARY KEY, b INT REFERENCES t1(b) ON UPDATE CASCADE ON UPDATE CASCADE);

Ordinarily I have to refer to UNIQUE columns, because it
should not be possible to have one foreign-key site that
matches two referenced-key sites. But using the
strange "affinity" rules, I can cause that situation.
These are weird results, but if there's
no plan to fix the data type assignment/comparison rules, I can
understand if you say "this is deliberate".
DROP TABLE t2;
DROP TABLE t1;
CREATE TABLE t1(a INT PRIMARY KEY, b CHAR(5) UNIQUE);
CREATE TABLE t2(a INT PRIMARY KEY, b int REFERENCES t1(b) ON UPDATE CASCADE);
INSERT INTO t1 VALUES (0, '2');
INSERT INTO t2 VALUES (0, '2');
INSERT INTO t1 VALUES (1, '02');
-- The following select will say there are 2 matching rows
SELECT COUNT(*) FROM t1 WHERE b IN (SELECT b FROM t2);

**Comment by Korablev77** [comment]
As for first point: I haven't dived into ANSI, but this behaviour looks satisfying (at least for me). Lets ask
other members for their opinion.

As far as I remember, `REPLACE` causes `DELETE + INSERT` on purpose. Again, personally I see no problem in such behaviour.

Currently, we have following grammar for constraints definition (I simplified it a little bit):
```
FK constraint ::= REFERENCES nm(T) eidlist_opt(TA) refargs(R)
refargs(A) ::= .
refargs(A) ::= refargs(A) refarg(Y) .
refarg(A) ::= ON INSERT | ON UPDATE | ON DELETE refact .
refact(A) ::= SET NULL | SET DEFAULT | SET CASCADE | RESTRICT | NO ACTION .
```
One can see that your example fits into this grammar. So, I guess if someday we decided to rework parser, we would fix grammar.
